File: 7-reverse-integer/7-reverse-integer.py
class Solution(object):
    def reverse(self, x):
        """
        :type x: int
        :rtype: int
        """
        MAX_32_VALUE = 2147483648
        REVERSED_MAX_32_VALUE = 463847412
        
        # Check negative 
        isNegative = False
        if x < 0:
            x = -x
            isNegative = True
            
        # Check if reversed is outside of [-2**31, 2**31 - 1]
        if x > 10**9:
            for i in range(1, 10):
                xMod10i = x % (10**i)
                reversedMax = REVERSED_MAX_32_VALUE % (10**i)
                print("i:" + str(i) + "->" + str(xMod10i) + " " + str(reversedMax))
                if xMod10i > reversedMax:
                    return 0
                elif xMod10i < reversedMax:
                    break
            
        
        # Reverse integer
        reversedInt = 0
        while x > 0:
            reversedInt *= 10
            reversedInt += x % 10
            x = int(x / 10)
            
        return -reversedInt if isNegative else reversedInt

File: 7-reverse-integer/7-reverse-integer.cs
public class Solution {
    const int INT_MAX = 2147483647;
    public int ReverseHelper(int x, int reversed) {
        if (reversed > (int) (INT_MAX / 10)) {
            return 0;
        }
        
        if (x < 10) {
            return reversed * 10 + x;
        }
        
        return ReverseHelper(x / 10, reversed * 10 + x % 10);
    }
    
    public int Reverse(int x) {
        if (x == - INT_MAX - 1) {
            return 0;
        }
        else if (x < 0) {
            return -1 * ReverseHelper(Math.Abs(x), 0);
        }
        return ReverseHelper(Math.Abs(x), 0);
    }
}